src/assert.js

```diff
@@ -429,61 +390,80 @@ class Assert {
+					expected: data.expected,
+					message: data.message
+				} );
+				done();
+			}
+		);
+	}
+}
 
-					// Expected is a constructor, maybe an Error constructor
-				} else if ( expectedType === "function" && actual instanceof expected ) {
-					result = true;
+function validateException( actual, expected, message, currentTest, assertionMethod ) {
+	let result = false;
+	const expectedType = objectType( expected );
 
-					// Expected is an Error object
-				} else if ( expectedType === "object" ) {
-					result = actual instanceof expected.constructor &&
-						actual.name === expected.name &&
-						actual.message === expected.message;
+	// We don't want to validate
+	if ( expected === undefined || expected === null ) {
+		result = true;
 
-					// Log the string form of the Error object
-					expected = errorString( expected );
+	// Expected is a regexp
+	} else if ( expectedType === "regexp" ) {
+		result = expected.test( errorString( actual ) );
 
-					// Expected is a validation function which returns true if validation passed
-				} else {
-					if ( expectedType === "function" ) {
-						result = expected.call( {}, actual ) === true;
-						expected = null;
+		// Log the string form of the regexp
+		expected = String( expected );
 
-						// Expected is some other invalid type
-					} else {
-						result = false;
-						message = "invalid expected value provided to `assert.rejects` " +
-							"callback in \"" + currentTest.testName + "\": " +
-							expectedType + ".";
-					}
-				}
+	// Expected is a constructor, maybe an Error constructor.
+	// Note the extra check on its prototype - this is an implicit
+	// requirement of "instanceof", else it will throw a TypeError.
+	} else if ( expectedType === "function" &&
+		expected.prototype !== undefined && actual instanceof expected ) {
+		result = true;
 
+	// Expected is an Error object
+	} else if ( expectedType === "object" ) {
+		result = actual instanceof expected.constructor &&
+			actual.name === expected.name &&
+			actual.message === expected.message;
 
-				currentTest.assert.pushResult( {
-					result,
+		// Log the string form of the Error object
+		expected = errorString( expected );
 
-					// leave rejection value of undefined as-is
-					actual: actual && errorString( actual ),
-					expected,
-					message
-				} );
+	// Expected is a validation function which returns true if validation passed
+	} else if ( expectedType === "function" ) {
 
-				done();
-			}
-		);
+		// protect against accidental semantics which could hard error in the test
+		try {
+			result = expected.call( {}, actual ) === true;
+			expected = null;
+		} catch ( e ) {
+
+			// assign the "expected" to a nice error string to communicate the local failure to the user
+			expected = errorString( e );
+		}
+
+	// Expected is some other invalid type
+	} else {
+		result = false;
+		message = "invalid expected value provided to `assert." + assertionMethod + "` " +
```

Comment:
Is it possible to reach this deterministically without the `actual` value (before the callback is invoked)? If so, we may want to `throw` hard here instead of pushing a new/alternate assertion, as we do with other validation errors (such as when a string is passed, earlier on in the `assert.throws()` code).

Reply:
There's one branch checking both:
```js
	} else if ( expectedType === "function" &&
		expected.prototype !== undefined && actual instanceof expected ) {
		result = true;
```
That gives a different "fallthrough" approach than if we just check if it's a function and then calculate the result in the body, because this one comes later:
```js
// Expected is a validation function which returns true if validation passed
} else if ( expectedType === "function" ) {
```

But... that means we can still check if it's a function and know that it wouldn't make it to the "invalid" case. So yes, we can validate the expected before doing anything with the actual!

I can make that change independent of what it does for that case: fail or error. Are you suggesting it should now error? The change would affect code like
```js
assert.throws(fcn, 2)
```
that would fail anyway. Erroring would be really clean and probably better if it's going _ahead_ of the invocation/resolution.

That was verbose, but thinking aloud; just want to make sure I have that all right. I'm onboard with such a change, even in this PR (creeping scope a bit, but it's all cohesive).

Reply:
Yeah, I think it'd be good to have the validation error be deterministic and thrown early as an exception rather than assertion failure.